### PR TITLE
Initial support for poll-based backend

### LINKED ISSUE #1604: Support for using poll as a backend
Tracking issue for #1602

See PR for a full technical description.

---

This is tracking the integration of a `mio` backend using the [poll(2)](https://man7.org/linux/man-pages/man2/poll.2.html) syscall for Unix like OS that don't support `epoll` or `kqueue`.

Technically this can target every OS which supports the `poll` (or even `select` syscall).

### PR DESCRIPTION
**Pull request by jasta**
With `mio_unsupported_force_poll_poll`, mio will use the legacy `poll(2)` syscall instead of `epoll` or `kqueue`.  While this is not intended to replace either of those more modern and efficient alternatives, it is a suitable avenue to enable a lot of other somewhat more obscure platforms such as the esp32 family or Haiku OS.  esp-idf support is my primary purpose and I believe the esp-rs community can get behind this as a way to better interop with a large number of Tokio-based async crates out there, but this PR is intended only to provide the barebones technical support for now with minimal invasion into existing production code paths.

This work is largely based off #1602 with improvements to fix tests and provide a more consistent implementation.  Specific changes include:

1. Closed fds are now automatically removed as is the case with the Windows implementation
2. `Waker` is now a pass through for the internal waker that the poll Selector needs
4. Haiku and esp-idf specific changes were removed.  IMO, those should be introduced as part of separate PRs for each respective platform.

Closes #1604

### CODE REVIEW

File: tests/tcp_stream.rs [hunk 1]
```diff
@@ -647,6 +647,7 @@ fn tcp_reset_close_event() {
     loop {
         poll.poll(&mut events, Some(Duration::from_millis(100)))
             .expect("poll failed");
+        println!("Got: {events:?}");
```
Comment (jasta) [inline tests/tcp_stream.rs:650]: Oops, I'll tidy up these leftovers

File: tests/unix_stream.rs [hunk 2]
```diff
@@ -451,6 +451,8 @@ where
 
     assert!(stream.take_error().unwrap().is_none());
 
+    assert_would_block(stream.read(&mut buf));
```
Comment (jasta) [inline tests/unix_stream.rs:454]: This is the same behaviour that tcp_stream has and is functionally important for the edge triggered emulation to work.  Basically if you only read exactly the data you need then nothing would trigger the EAGAIN case and the fd wouldn't be re-registered with POLLIN.  Technically this doesn't feel like a big issue in practice though since in any real world application you'd have to read all available buffered data up to EAGAIN to know for sure that subsequent calls to epoll would actually tell you if data was ready.  There could be contrived and very bizarre usages though that would break...

**Comment by jasta** [comment]
Also, I wonder if maybe the WakerRegistrar thing would be better served by copying the kqueue design pattern whereby the Selector provides factory-type methods to create and use a waker.  I didn't notice that it already is working pretty similarly to what I need until just now...

**Comment by jasta** [comment]
Indeed, I think I found a much cleaner way to detangle Waker behaviour from the Selector so that we can have the concept of a WakerDriver (a low-level description of how to call eventfd, pipe, pipe2, or kevent) vs a Waker which is created by the selector and introduces some customization of the WakerDriver behaviour _based on_ the selector's quirks.  For example, the poll selector can use either eventfd or pipe2 WakerDriver, but the Waker it creates requires reregistering the interest with poll before wake can be called.  The epoll selector can use either eventfd or pipe2 WakerDriver, but the Waker it creates has a no-op before each call to wake.  

This seems to elegantly resolve one existing gotcha and two that I created:

1. pipe::Waker::wake no longer needs an illumos-specific cfg check.  Instead, the epoll selector will have that check since it is epoll-specific behaviour on that platform that's the issue.
2. I can drop my WakerRegistrar as I can now have the poll selector directly control reregistration of whatever WakerDriver is used
3. The poll selector won't need another eventfd/pipe impl living in side the poll selector for the purposes of internally waking a blocking poll call.  We can just use the WakerDriver internally.

This also fixes the FreeBSD failing tests (which are actually just because WakerRegistrar is unused in the kqueue path)

File: src/sys/unix/selector/adapters/level_triggered/io_source_state.rs [hunk 3]
```diff
@@ -0,0 +1,106 @@
+use crate::sys::Selector;
+use crate::{Interest, Registry, Token};
+use std::io;
+use std::os::unix::io::RawFd;
+use std::sync::Arc;
+use crate::sys::unix::selector::poll::RegistrationRecord;
+
+struct InternalState {
```
Comment (Thomasdezeeuw) [inline src/sys/unix/selector/adapters/level_triggered/io_source_state.rs:8]: This is a lot of state per fd.

The wasi implementation is based of Posix `poll` (Wasi's `poll_oneoff`) and doesn't need to hold any state, but requires proper (de)registering, can we follow that implementation more closely?

**Comment by jasta** [comment]
Oh and also I could break up the refactor into a separate PR so it's easier to swallow and confirm that it's correct without introducing the poll backend.

File: src/sys/unix/selector/adapters/level_triggered/io_source_state.rs (hunk 3 above)
Comment (jasta) [reply src/sys/unix/selector/adapters/level_triggered/io_source_state.rs:8]: This is basically a copy of how the Windows impl does it, and was the same in #1602.  I don't think the wasi approach will work, and actually I'm not really sure the wasi implementation is event correct.  It's certainly missing a lot of features and doesn't behave the same as epoll/kqueue.  The poll behaviour in this PR is trying to be as close as possible to a correct epoll emulation with only one leak that I'm aware of (see the change to tcp_stream test)

**Comment by jasta** [comment]
I was wrong re my statement that any UNIX should be able to support the poll+pipe backend.  While technically true that it will "work", it's hopeless to get all current tests to pass because libc::POLLRDHUP is Linux-specific and without it tests like tests/tcp.rs:write_shutdown will fail.  There are about a dozen tests like this.  This does mean that mio still works of course, just that shutdown(Read/Write) will not behave as expected and the caller will need to detect these shutdown events by trying to read or write.

**Comment by Thomasdezeeuw** [comment]
I wanted to review this in the weekend, but I didn't. I think this still has a lot of changes that are not strictly necessary for `poll(2)` support. From the changes to `IoSourceState` (can be contained to `src/sys/unix/poll.rs`), to the macros (not required), but also the waker (this should just be an pipe/eventfd fd with `IoSourceState`). Also unnecessary name changes `Waker` to `WakerDriver`?

Maybe I have time next weekend, but not promises, I do want to reduce the number of changes though. I really like to see a minimal set of changes to implement this.

**Comment by jasta** [comment]
> From the changes to `IoSourceState` (can be contained to `src/sys/unix/poll.rs`)

Ackd, I'll avoid breaking things out into separate files.

> ...to the macros (not required)

I only did this because the original PR reviewer asked for it because the PR will require copy/pasting the cfg criteria quite a few times.  I can drop it though, but see my comments below so we're on the same page as to how many times this will need to get repeated.

> ...but also the waker (this should just be an pipe/eventfd fd with `IoSourceState`)

I'm not sure how this can be achieved.  The critical difference between epoll vs poll is that for poll you either need to read the pipe/eventfd when it wakes (what the original PR did) _or_ you need to reregister the interests again before calling wake (what my PR did).  In either case, there needs to be code that allows the poll Selector to interact with Waker internals.  IoSourceState is also but separately interacting with Selector internals.

Fwiw the reason I opted for a different strategy was so that I didn't need to deepen the circular relationship between Selector and Waker but could rather describe the dependencies as such:

```
Selector -> Waker -> WakerDriver
Waker -> Selector
```

>Also unnecessary name changes `Waker` to `WakerDriver`?

There wasn't a rename here as such, but rather WakerDriver was added to be used by Waker.  This is because as I mentioned above in a previous comment, we need to use pipe or eventfd to internally wake poll() when register() is called.  But Waker as written right now calls register in the constructor making it a bit fragile to reuse internally.  So I stripped WakerDriver away which is just a pure side effect free API and then added Waker back inside the Selector with the register side effect.  This allowed the poll Selector to internally use WakerDriver and this neither duplicated code nor, as the original PR did, unintentionally use pipe for the internal register() wakeup case but eventfd for the Waker case.

> 
> Maybe I have time next weekend, but not promises, I do want to reduce the number of changes though. I really like to see a minimal set of changes to implement this.

Understood, I'm willing to rework the PR with reduced churn, it's just not 100% clear where the tradeoffs should be.  See my comment above about how clumsy it gets to try to conditionally call reregister() inside wake() only for the poll Selector case.  If that clumsiness is preferred, that's what I'll do.  

It is, I admit, less intrusive to use the original PR's strategy for reading the pipe/eventfd on wake tho since you only need to add new constructor APIs that don't do the register (e.g. new_internal), then add a new clear() API that resets them as needed.  I'll play around with this approach some more and see if I can't get it to look a little less clumsy than the original PR had it.

**Comment by jasta** [comment]
Also wanted to thank you for taking the time to review and discuss. I know this poll support is probably weighing on you as a reviewer a bit after so much churn and iteration.  This will be such a huge help for the esp32 community once we're able to get it over the line though!

**Comment by Thomasdezeeuw** [comment]
> > ...to the macros (not required)
> 
> I only did this because the original PR reviewer asked for it because the PR will require copy/pasting the cfg criteria quite a few times. I can drop it though, but see my comments below so we're on the same page as to how many times this will need to get repeated.

I'm personally a bit torn on the macros. For things like `cfg_net` they help, but for the selector I'm no so sure, I think the code/`cfg`s is easier to follow without th emacro.

> > ...but also the waker (this should just be an pipe/eventfd fd with `IoSourceState`)
> 
> I'm not sure how this can be achieved. The critical difference between epoll vs poll is that for poll you either need to read the pipe/eventfd when it wakes (what the original PR did) _or_ you need to reregister the interests again before calling wake (what my PR did). In either case, there needs to be code that allows the poll Selector to interact with Waker internals. IoSourceState is also but separately interacting with Selector internals.

I was thinking an eventfd/pipe who's fd is always passed to poll (assuming the waker was added), preferably in a fixed position in the `pollfds`. Then we always check the waker's `pollfd` and drain it's readiness if we get an event (e.g. read from the pipe). For the `Waker` we can't use the kqueue implementation as it has no fd, but the eventfd and pipe should continue working as normal, I would expect. The only change in the `Waker` code would be a special call to `register`, something like `selector.register_waker`, which for `poll(2)` would mean to duplicate the fd and use it as described above. For epoll we can just call `register.register` (as we currently do).

> Fwiw the reason I opted for a different strategy was so that I didn't need to deepen the circular relationship between Selector and Waker but could rather describe the dependencies as such:
> 
> ```
> Selector -> Waker -> WakerDriver
> Waker -> Selector
> ```
> 
> > Also unnecessary name changes `Waker` to `WakerDriver`?
> 
> There wasn't a rename here as such, but rather WakerDriver was added to be used by Waker. This is because as I mentioned above in a previous comment, we need to use pipe or eventfd to internally wake poll() when register() is called. But Waker as written right now calls register in the constructor making it a bit fragile to reuse internally. So I stripped WakerDriver away which is just a pure side effect free API and then added Waker back inside the Selector with the register side effect. This allowed the poll Selector to internally use WakerDriver and this neither duplicated code nor, as the original PR did, unintentionally use pipe for the internal register() wakeup case but eventfd for the Waker case.

That makes more sense. If my idea describe above doesn't work we can go with something like this. But I'm not sure about the "driver" name, but we can bikeshed this later.

> > Maybe I have time next weekend, but not promises, I do want to reduce the number of changes though. I really like to see a minimal set of changes to implement this.
> 
> Understood, I'm willing to rework the PR with reduced churn, it's just not 100% clear where the tradeoffs should be. See my comment above about how clumsy it gets to try to conditionally call reregister() inside wake() only for the poll Selector case. If that clumsiness is preferred, that's what I'll do.

See above, but I prefer to limit the scope of this pr, affecting the minimal amount of non-poll(2) code would be ideal. That makes it easier to review and less risky to merge as other code isn't affected.

> It is, I admit, less intrusive to use the original PR's strategy for reading the pipe/eventfd on wake tho since you only need to add new constructor APIs that don't do the register (e.g. new_internal), then add a new clear() API that resets them as needed. I'll play around with this approach some more and see if I can't get it to look a little less clumsy than the original PR had it.



> Also wanted to thank you for taking the time to review and discuss. I know this poll support is probably weighing on you as a reviewer a bit after so much churn and iteration. This will be such a huge help for the esp32 community once we're able to get it over the line though!

I appreciate this. I very much want to help the esp32 community, but my main priority is to keep the existing code working and maintainable (for my own sanity).

**Comment by jasta** [comment]
> I was thinking an eventfd/pipe who's fd is always passed to poll (assuming the waker was added), preferably in a fixed position in the pollfds. Then we always check the waker's pollfd and drain it's readiness if we get an event (e.g. read from the pipe). For the Waker we can't use the kqueue implementation as it has no fd, but the eventfd and pipe should continue working as normal, I would expect. The only change in the Waker code would be a special call to register, something like selector.register_waker, which for poll(2) would mean to duplicate the fd and use it as described above. For epoll we can just call register.register (as we currently do).

Looking into more detail here and this is basically how the original PR (#1602) worked but it needed unsafe handling of the waker RawFd and a change to all sys::Selector impls.  I also realized that your suggestion to dup the fd won't work directly directly because esp32 doesn't implement dup _but_ it can work indirectly by just internally re-using the waker we know the poll-impl already has:

```rust
#[cfg(mio_unsupported_force_poll_poll)]
mod poll {
  pub struct Waker {
    selector: sys::Selector,
  }

  impl Waker {
    pub fn new(selector: &Selector, token: Token) -> io::Result<Waker> {
      Ok(Waker { selector: selector.try_clone()? })
    }

    pub fn wake(&self) -> io::Result<()> {
      // Since this is the poll-only impl, the Selector would need only to call self.notify_waker.wake() and everything would
      // just work the exact same way it does when you call Selector::register.
      select.selector.wake_internal()
    }
  }
}
```

This would then save on an fd and make the code path far, far clearer how it's being used.  So, the only real issue then is how can we get the poll impl to re-use the code we have in eventfd::Waker and pipe::Waker.  I'd propose that once again the WakerDriver idea helps us here (pending a rename perhaps):

1. Expose pub(crate) versions of eventfd::Waker and pipe::Waker as a new type (for argument let's say WakerImpl) that doesn't interact with Selector at all (like what my PR did with WakerDriver).  This can be a simple rename mostly, but unlike my PR today will not include other Waker implementations and is only meant to be an internal sys/unix utility to re-use eventfd/poll impl code.
2. Convert the eventfd::Waker and pipe::Waker types to be simple pass-throughs to eventfd::WakerImpl and pipe::WakerImpl with the extra register behaviour on top.
3. Add a new poll::Waker as written above that is a simple pass-through to `poll::Selector::wake_internal`.

Thoughts?  I'm working on this locally now in parallel but I want to be respectful of your suggestions and not blindside you with changes to the PR that aren't quite like we discussed already.

**Comment by jasta** [comment]
Force pushed a reworked version of the PR that I think is much cleaner based on your feedback and on my comment above.  Heads up that there is a functional change here to re-use the internal Waker which both simplifies the implementation and makes it much clearer why we had to slightly refactor eventfd::Waker and pipe::Waker.

**Comment by jasta** [comment]
Silly mistake had the CI checks not running, now that they're up and failing I'll go ahead and fix.

File: src/sys/unix/selector/poll.rs [hunk 4]
```diff
@@ -0,0 +1,725 @@
...
+            self.operations_complete.notify_one();
+        }
+
+        res
+    }
+
+    /// Special optimized version of [Self::deregister] which handles multiple removals
+    /// at once.  Ok result if all removals were performed, Err if any entries
+    /// were not found.
+    fn deregister_all(&self, targets: &[RawFd]) -> Result<(), ()> {
```
Comment (Thomasdezeeuw) [inline src/sys/unix/selector/poll.rs:411]: I don't think this function, as used by `SelectorState::select`, should return an error, it should just ignore them and continue removing all the fds. Only the single `deregister` should return an error.

File: src/sys/unix/selector/poll.rs [hunk 5]
```diff
@@ -0,0 +1,725 @@
...
+        // <poll happens>
+        //
+        // Fd's pending removal only get cleared when poll has been run. It is possible that
+        // between registering and deregistering and then _again_ registering the file descriptor
+        // poll never gets called, thus the fd stays stuck in the pending removal list.
+        //
+        // To avoid this scenario we remove an fd from pending removals when registering it.
+        let mut pending_removal = self.pending_removal.lock().unwrap();
+        if let Some(idx) = pending_removal.iter().position(|&pending| pending == fd) {
+            pending_removal.remove(idx);
```
Comment (Thomasdezeeuw) [inline src/sys/unix/selector/poll.rs:332]: I don't think we care about the order, right?
```suggestion
            pending_removal.swap_remove(idx);
```

File: src/sys/unix/selector/poll.rs [hunk 6]
```diff
@@ -0,0 +1,725 @@
...
+        drop(pending_removal);
+
+        self.modify_fds(|fds| {
+            if fds.fd_data.contains_key(&fd) {
+                return Err(io::Error::new(
+                    io::ErrorKind::AlreadyExists,
+                    "\
+                    same file descriptor registered twice for polling \
+                    (an old file descriptor might have been closed without deregistration)\
+                    ",
```
Comment (Thomasdezeeuw) [inline src/sys/unix/selector/poll.rs:343]: Nit: to match some of the other related error message.
```suggestion
                    "I/O source already registered this `Registry` \
                    (an old file descriptor might have been closed without deregistration)",
```

File: src/sys/unix/selector/poll.rs [hunk 7]
```diff
@@ -0,0 +1,725 @@
...
+        let timeout_ms = deadline
+            .map(|deadline| {
+                let timeout = deadline.saturating_duration_since(Instant::now());
+
+                // Round up to a whole millisecond.
+                let mut ms = timeout.as_millis().try_into().unwrap_or(u64::MAX);
+                if Duration::from_millis(ms) < timeout {
+                    ms = ms.saturating_add(1);
+                }
+                ms.try_into().unwrap_or(i32::MAX)
```
Comment (Thomasdezeeuw) [inline src/sys/unix/selector/poll.rs:516]: Can you use the same code as we use for the `epoll(2)` implementation: https://github.com/tokio-rs/mio/blob/2123bedbed00e24dccc036a0e1799e9e1d7dba1c/src/sys/unix/selector/epoll.rs#L90-L98

This way we have consistent timeout rounding.

File: src/sys/unix/selector/poll.rs [hunk 8]
```diff
@@ -0,0 +1,725 @@
...
+            libc::POLLWRBAND,
+            libc::POLLERR,
+            libc::POLLHUP,
+        );
+
+        // Can't reference fields in packed structures.
+        let e_u64 = event.token.0;
+        f.debug_struct("epoll_event")
+            .field("events", &EventsDetails(event.events))
+            .field("u64", &e_u64)
```
Comment (Thomasdezeeuw) [inline src/sys/unix/selector/poll.rs:618]: Little copy and paste fix ;)
```suggestion
        f.debug_struct("poll_event")
            .field("token", &event.token)
            .field("events", &EventsDetails(event.events))
```

File: src/sys/unix/selector/poll.rs [hunk 9]
```diff
@@ -0,0 +1,725 @@
...
+
+            if let Err(err) = &result {
+            if err.kind() == io::ErrorKind::WouldBlock {
+                self.inner.as_ref().map_or(Ok(()), |state| {
+                    state
+                    .selector
+                    .reregister(state.fd, state.token, state.interests)
+                })?;
+            }
+        }
```
Comment (Thomasdezeeuw) [inline src/sys/unix/selector/poll.rs:665]: No automatic formatting within macro :(
```suggestion
                if err.kind() == io::ErrorKind::WouldBlock {
                    self.inner.as_ref().map_or(Ok(()), |state| {
                        state
                        .selector
                        .reregister(state.fd, state.token, state.interests)
                    })?;
                }
            }
```

**Review by Thomasdezeeuw — COMMENTED** [review]
I finally got some time to review this in its entirety. 

I got a couple of small comment, some tiny nits, but overall this LGTM.

File: .github/workflows/ci.yml [hunk 10]
```diff
@@ -31,6 +31,18 @@ jobs:
       run: cargo test --all-features
     - name: Tests release build
       run: cargo test --release --all-features
+  TestPoll:
+    runs-on: ubuntu-latest
```
[the remark this answers is not in the document]
Comment (jasta) [reply .github/workflows/ci.yml:35]: Unfortunately, no.  See a previous comment calling out that without SIGRDHUP, which is Linux only, several tests that check for is_read_closed will fail.  I can conditionally disable those tests but I don't think they can be fixed.

File: src/sys/unix/selector/poll.rs [hunk 11]
```diff
@@ -0,0 +1,725 @@
...
+                        // Store event
+                        events.push(Event {
+                            token: fd_data.token,
+                            events: poll_fd.revents,
+                        });
+
+                        if poll_fd.revents & (libc::POLLHUP | libc::POLLERR) != 0 {
+                            pending_removal.push(poll_fd.fd);
+                            closed_raw_fds.push(poll_fd.fd);
+                        }
```
[the remark this answers is not in the document]
Comment (jasta) [reply src/sys/unix/selector/poll.rs:280]: So there's a few ways to do this but at a high level the problem to solve is that with edge triggered behavior once an fd is closed or shutdown you won't get spammed with err/hup events in a tight loop; with level triggered you will.  So once you receive a terminal event from poll you have to internally track that you shouldn't call poll anymore with that fd (removing events flags isn't enough) but let external callers still safely call deregister since ofc from their POV it is still registered.

In effect the solution I went with (which is also how the Windows Selector seems to do it) is that but it turns deregister into a two layered concept: the lowest layer removes the fd entirely and just forgets it exists, while the higher layer uses an AtomicBoolean to stay aligned with the lower layer and no op calls to deregister.

There are alternatives though, such as simply removing the fds from the list of fds sent to poll but still keeping FdData around with a new flag indicating it's closed.  Then when you deregister it just ignores that it isn't in the poll_fds set.  I like this solution generally but chose not to go there as I think it's slightly more fragile as there's already quite a few edge cases in the poll loop to keep track of.  Additionally there was an optimization I avoided (you can modify the internal data structures while holding the lock inside the main poll loop) simply to avoid making that code even harder to follow / more likely to break.

I'm interested in feedback here though and happy to pivot, just wanted to help you understand my thinking so far.

File: src/sys/unix/selector/poll.rs [hunk 12]
```diff
@@ -0,0 +1,725 @@
...
+
+                        // Remove the interest which just got triggered
+                        // the IoSourceState/WakerRegistrar used with this selector will add back
+                        // the interest using reregister.
+                        poll_fd.events &= !poll_fd.revents;
+
+                        if events.len() == num_fd_events {
+                            break;
+                        }
```
[the remark this answers is not in the document]
Comment (jasta) [reply src/sys/unix/selector/poll.rs:289]: This is part of the code that was lifted with permission and I'm not 100% sure why the author did this.  Best I can tell it is a slight optimization where if say you tracked 100 fds but the first one was the only one with revents you could avoid iterating over the other 99.  I'm on my mobile and can't easily confirm if the order is insert order or random but in either case I doubt this optimization changes the average runtime in an academic sense so probably can just be dropped.

File: src/sys/unix/selector/poll.rs [hunk 13]
```diff
@@ -0,0 +1,718 @@
...
+/// Unique id for use as `SelectorId`.
+#[cfg(debug_assertions)]
+static NEXT_ID: AtomicUsize = AtomicUsize::new(1);
+
+#[derive(Debug)]
+pub struct Selector {
+    state: Arc<SelectorState>,
+    /// Whether this selector currently has an associated waker.
+    #[cfg(debug_assertions)]
+    has_waker: AtomicBool,
```
[the remark this answers is not in the document]
Comment (jasta) [reply src/sys/unix/selector/poll.rs:29]: Not sure I fully grok the relationship between Registry and Selector and SelectorState (I understood it was 1:1, so it shouldn't matter which struct the AtomicBool is in?), but in either case the poll Selector is similar to how the epoll and Windows Selector works, though I do detect a small bug in poll::Selector::try_clone() which resets has_waker to false (which I'll push a fix for)

File: src/sys/unix/selector/poll.rs [hunk 14]
```diff
@@ -0,0 +1,725 @@
...
+        })
+    }
+
+    pub fn select(&self, events: &mut Events, timeout: Option<Duration>) -> io::Result<()> {
+        let deadline = timeout.map(|t| Instant::now() + t);
+
+        events.clear();
+
+        let mut fds = self.fds.lock().unwrap();
+        let mut closed_raw_fds = Vec::new();
```
[the remark this answers is not in the document]
Comment (jasta) [reply src/sys/unix/selector/poll.rs:202]: Yes, but not in a way that's all that straight forward to test.  I'm interested in this optimization as today we allocate at least once each time an fd is closed, but I avoided a potential dead lock case by trying to apply the removals while holding the fds lock.  The most efficient solution is to then move closed_raw_fds inside the fds Mutex and apply the removals while still holding the lock, but this isn't straight forward to test that we didn't introduce subtle dead locks so I'm really trying to punt here if we can.

If you feel strongly though I'll rework, though I'll really need to think hard about how to create an effective test to prevent the deadlock.

**Comment by jasta** [comment]
Applied most of the suggested changes (see comments above on the others) and confirmed tests still pass.  If you prefer I can squash all commits to prepare to merge.

File: src/sys/unix/selector/poll.rs [hunk 15]
```diff
@@ -0,0 +1,725 @@
...
+            }
+
+            trace!("Poll finished: {:?}", &fds);
+            let notified_events = fds.poll_fds[0].0.revents;
+            let notified = notified_events != 0;
+            let mut num_fd_events = if notified { num_events - 1 } else { num_events };
+
+            let mut pending_wake_tokens_guard = self.pending_wake_tokens.lock().unwrap();
+            let pending_wake_tokens =
+                std::mem::replace(pending_wake_tokens_guard.as_mut(), Vec::new());
```
[the remark this answers is not in the document]
Comment (jasta) [reply src/sys/unix/selector/poll.rs]: Same as above, I'm worried about how prone this code is to dead locks.  Ofc it's possible to fix, just hard to create robust tests for the edge cases

File: src/sys/unix/selector/poll.rs [hunk 16]
```diff
@@ -0,0 +1,725 @@
...
+
+            if notified {
+                self.notify_waker.ack_and_reset();
+                num_fd_events += pending_wake_tokens.len();
+            }
+
+            // We now check whether this poll was performed with descriptors which were pending
+            // for removal and filter out any matching.
+            let mut pending_removal_guard = self.pending_removal.lock().unwrap();
+            let mut pending_removal = std::mem::replace(pending_removal_guard.as_mut(), Vec::new());
```
[the remark this answers is not in the document]
Comment (jasta) [reply src/sys/unix/selector/poll.rs:239]: Same as above, I'm worried about how prone this code is to dead locks. Ofc it's possible to fix, just hard to create robust tests for the edge cases

File: src/sys/unix/selector/poll.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (jasta) [reply src/sys/unix/selector/poll.rs]: Fixed by using Option

File: src/sys/unix/selector/poll.rs (hunk 4 above)
Comment (jasta) [reply src/sys/unix/selector/poll.rs:404]: It returns so that deregister() can reuse the code and forward the error from it

Comment (Thomasdezeeuw) [reply src/sys/unix/selector/poll.rs:405]: I understand that, but on line 416 it does the following:

```rust
            for target in targets {
                let data = fds.fd_data.remove(&target).ok_or(())?;
```
So that the first error it hits will be returned, not removing the other fds. For a single `deregister` call by the user we want this, but not in `Selector::select`, here we want to continue to remove the other fds as well (we ignore any errors from `deregister_all` any way).

File: src/sys/unix/selector/poll.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (Thomasdezeeuw) [reply src/sys/unix/selector/poll.rs:283]: I see, well maybe it's worth a comment saying why it's in there, but otherwise you can leave it in.

File: src/sys/unix/selector/poll.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (Thomasdezeeuw) [reply src/sys/unix/selector/poll.rs:274]: That makes sense... I don't really have a good intuition into what option would be best, so instead let me throw another idea in the ring.

Instead of removing the fd, we can also set the fd to a negative number (i.e. `poll_fd.fd -= poll_fd.fd`), then it will be ignored by the kernel. A little less overhead for us, but a little more for the kernel...

But as I said, I'm not sure what the best option is.

I'm fine with any of them really, they all make sense to me, but a comment explaining the situation would be nice.

File: src/sys/unix/selector/poll.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (Thomasdezeeuw) [reply src/sys/unix/selector/poll.rs:239]: I'm not saying we add any locks or keep the lock for longer. I'm say instead of using `Vec::new` we add that vector to `Selector` (not `SelectorState`) and do `swap(pending_removal_guard, self.pending_removal_buffer)`, this way we reuse the allocation and we can keep processing it without holding the lock.

File: src/sys/unix/selector/poll.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (Thomasdezeeuw) [reply src/sys/unix/selector/poll.rs:202]: We can pass a `close_raw_fds` vector to the function and keep it in `Selector` (not `SelectorState`).

[the remark this answers is not in the document]
Comment (Thomasdezeeuw) [reply src/sys/unix/selector/poll.rs:202]: Oh btw, `Selector::select` can take a mutable reference to self (`&mut self`), so anything in `Selector` can be freely accessed. Maybe that wasn't clear as most Unix implementation don't make use of that.

File: src/sys/unix/selector/poll.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (Thomasdezeeuw) [reply src/sys/unix/selector/poll.rs:239]: Also see my comment about `Selector::selector` taking a mutable self reference.

File: .github/workflows/ci.yml (hunk 10 above)
[the remark this answers is not in the document]
Comment (Thomasdezeeuw) [reply .github/workflows/ci.yml:35]: Ok. `Event::is_read_closed` is provided on an best effort basis, so it's acceptable if that doesn't trigger. Let's do this as a follow up then.

File: src/sys/unix/selector/poll.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (Thomasdezeeuw) [reply src/sys/unix/selector/poll.rs:28]: Okay, so all implementation have this problem. I've opened https://github.com/tokio-rs/mio/issues/1698. For now it's fine to leave it as is.

File: src/sys/unix/selector/poll.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (jasta) [reply src/sys/unix/selector/poll.rs:202]: Normally true, but the way the poll impl works is that there needs to be a non-exclusive relationship between Registry and Selector such that `Registry::register` will ultimately call into `poll::IoSourceState::register` which accesses the selector using `registry.selector()` and then calls `register_internal`.  This test proves that we cannot have exclusive access to `poll::Selector::poll`: https://github.com/tokio-rs/mio/blob/master/tests/poll.rs#L323 (i.e. we must be able to call `SelectorState::register_internal` and `SelectorState::select` simultaneously).  This is also why we've got a Mutex for accessing `poll::SelectorState::fds` (register_internal and select both want to modify this data structure).

And that's why the dead lock can be so insidious -- if you try to re-use the memory for `closed_raw_fds` then you will have to extend the lock's duration (or add another one and there's already 3!) in a way that will start to overlap the locks on top of each other making a dead lock opportunity considerably more likely.  I'm confident I can apply the optimization without a dead lock, but I'm not so confident I can devise a reliable test that will prevent a dead lock regression from easily slipping in later.

I vote that we punt on the optimization for now given this concern, though I can push harder down this path if you prefer.

File: src/sys/unix/selector/poll.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (jasta) [reply src/sys/unix/selector/poll.rs:274]: Ack'd, I'll add some language around why we're removing before we actually deregister from the outside.

File: src/sys/unix/selector/poll.rs (hunk 4 above)
Comment (jasta) [reply src/sys/unix/selector/poll.rs:405]: Oh good catch and I misunderstood what you meant by return in that sense.  Yeah, I'll fix it.

File: src/sys/unix/selector/poll.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (Thomasdezeeuw) [reply src/sys/unix/selector/poll.rs:202]: I'm fine with punting this for now, we need to make a follow up issue any way (mostly documenting the implementation once we're ready to stabilise it). However I think you're misunderstand my suggestion.

I'm suggesting we add a field `closed_raw_fds: Vec<RawFd>` to `Selector`, change `Selector::select` to take `&mut self` allowing to accept this new field exclusively. Change `SelectorState::select` (this function) to accept `closed_raw_fds: &mut Vec<RawFd>` as an argument and use that as a buffer, rather then creating a new `Vec` in each call to `select`.

[the remark this answers is not in the document]
Comment (jasta) [reply src/sys/unix/selector/poll.rs:207]: Oh I see what you mean.  SelectorState won't hold the buffer and so won't need `&mut self`, but `Selector` can hold the buffer _however_ this means that you'll be cloning the buffer for every register that you do (see `poll::IoSourceState::register`'s use of `try_clone()`.  This means that you'll actually allocate _more_ than you were before since right now you only allocate on socket close instead of socket creation.

[the remark this answers is not in the document]
Comment (jasta) [reply src/sys/unix/selector/poll.rs:207]: Using #1701 to track these optimizations

**Comment by jasta** [comment]
Ok applied a few more suggestions and left a tracking issue for the optimizations.  Again, I can squash all these commits if you prefer, just lmk how you want to proceed to merge.

**Comment by jasta** [comment]
Just caught an issue from CI that needs to be addressed, working on fixing up a bad assumption that IoSourceState in sys/shell/mod could assume non-Windows was UNIX which breaks for wasm32-wasi

**Review by Thomasdezeeuw — APPROVED** [review]

**Comment by Thomasdezeeuw** [comment]
Thanks @jasta for pulling this over the finish line!
